[rocprofiler-sdk][rocprofiler-register] Initial Attachment Support

meserve-amd opened the pull request:
## Motivation

This PR is all changes needed to support attachment to a user's process for profiling. 

A user should be able to use `rocprofv3 -p <pid>` with their usual command line parameters and see the same functionality as if they had started their app with rocprofv3.

This feature is enabled with `ROCPROFILER_REGISTER_ATTACHMENT_QUEUES_ENABLED=1`, which must be present when the target program is started. Without this variable, attachment will succeed but kernel tracing will not be available, even if specified. (I will add a warning about this in rocprofv3 and remove this aside when I do)

## Technical Details

The largest changes are around the rocprofiler-attach component, which handles the mechanics of attachment and interacting with the ptrace system library. In addition, rocprofiler-sdk will be loaded after the program starts, so another large chunk of changes are to support this. These are the rocprofiler-sdk-prestore library as well as the code changes in queue_controller.cpp and code_object.cpp.

A summary of expected code flow:

Prior to attachment, rocprofiler-register is responsible for loading the rocprofiler-sdk-prestore library which will add calls to the HSA API table to capture queue creation and executable freezing. 

1. User calls `rocprofv3 -p <pid>`
    - Translates all command line parameters into environment variables as usual
2. rocprofv3 calls rocprofv3_attach.py with pid
3. rocprofv3_attach calls rocprofiler-attach component
    - Uses ptrace to seize the target pid
    - Uses ptrace to poke in the environment variables into a buffer for use by target program
    - Identifies the entry point in rocprofiler-register for attachment and invokes it using ptrace
4. rocprofiler-attach calls rocprofiler-register attach entrypoint in target
    - Re-creates the environment of rocprofv3 by reading the environment buffer
    - Calls rocprofiler-tool to initialize rocprofiler
    - Replays all previous API registrations for rocprofiler
    - Provides entrypoints from rocprofiler-sdk-prestore to rocprofiler to replay creation of queues and executables

From here, rocprofiler-sdk runs as before. When detach is called, it follows the same flow to rocprofiler-register which invokes the tool library to finalize which generates output.

In addition to this overview, I've included file level comments for each file to help reviewers.

## Limitations

Currently, these are limitations that are part of this PR but we intend to fix in the future:

- Detach on signal
    - Straightforward, need to add signal handling to rocprofv3_attach.py
- Attach multiple times to the same PID
    - Requires some changes to the prestore library and register to support unloading
- Support always, not just with environment variable
    - This has implications for ALL rocm apps, even when attachment is never used
    - Need to do more impact evaluation here
- Additional information saved in prestore
    - HIP stream IDs, possibly more
- Allow other tool libraries to be invoked
- Automated testing
    - Setting up the attachment scenario correctly in testing is challenging, but will need to be done

## Test Plan

Extensive manual testing has been done on the development branch prior to this PR. We have provided this code to other teams and fixed many issues as a result.

Automated testing will be developed and added for code coverage.

## Test Result

We have tested attachment with bare programs, sglang, and vllm. Issues arising out of that testing have been resolved.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

meserve-amd committed:
attach: milestone: API tracing

- This change must be paired with a change in rocprofiler-register to
  fully function.
- API tracing works at this commit
- Queue tracing not supported yet

meserve-amd committed:
attach: milestone: API tracing

- This pairs with another commit in rocprofiler-sdk to fully
  function
- Add ptrace entry points for tool attachment
- API tracing works at this commit
- Queue tracing not supported yet

meserve-amd committed:
attach: cleanup and comments

itrowbri committed:
attach: Formatting and crash fixes

meserve-amd committed:
attach: cleanup

- Remove hardcode for loading of tool library
- Make invoke registration functions public again

meserve-amd committed:
attach: add attach duration

- Add option attach-duration-msec for attachment

itrowbri committed:
Formatting + sglang hang fix via signal handling

itrowbri committed:
Changed FATAL_IF to DFATAL_IF for scratch_memory due to persistent crash when iterating queues

meserve-amd committed:
attach: proxy queue first draft

- Adds ability to trace with queues during attachment
- Must be paired with updated rocprofiler-register

meserve-amd committed:
attach: proxy queue first draft

- Adds ability to trace with queues during attachment
- Must be paired with updated rocprofiler-sdk

itrowbri committed:
Allow null agents for scratch output

meserve-amd committed:
attach: improve queue library interface

- Significant changes to force exported interfaces back to C
- Fixes bug with unknown agents at attachment
- Code objects' names may still be incorrect

meserve-amd committed:
attach: add code_object support

- Kernel traces will now have names and all other information for launches
- Add capture of hsa_executable to the queue library
- Various logging improvements

meserve-amd committed:
attach: rename queue library to prestore

meserve-amd committed:
attach: prestore overhaul

- Must be paired with commit in rocprofiler-sdk

meserve-amd committed:
attach: prestore overhaul

- Must be paired with commit from rocprofiler-register
- Massive overhaul of code organization in prestore library
  - Separates registrations for different object types
  - Sets up future changes for initialization

meserve-amd committed:
attach: add dispatch table rework

- Register will load the prestore library and provide entrypoints to sdk

meserve-amd committed:
attach: formatting and cleanup

meserve-amd committed:
attach: revise dispatch table scheme

meserve-amd committed:
attach: add prestore dispatch table

- Removes linkage to prestore library from sdk

meserve-amd committed:
attach: cleanup

meserve-amd committed:
attach: formatting

meserve-amd committed:
attach: formatting

meserve-amd committed:
attach: fix input prompt not appearing

meserve-amd commented on projects/rocprofiler-register/source/lib/rocprofiler-register/rocprofiler_register.cpp:
Large changes here:
- Load prestore library when attachment is enabled
- Create dispatch table of prestore library for rocprofiler-sdk once it's loaded
- Control attachment by loading environment buffer and starting the tool

Minor:
- Made load lib more general

meserve-amd committed:
attach: fix component name in cmake

meserve-amd commented on projects/rocprofiler-sdk/cmake/rocprofiler_config_packaging.cmake:
Add rocprofiler-sdk-prestore component
Also fixed a typo in the error here

meserve-amd commented on projects/rocprofiler-sdk/source/bin/CMakeLists.txt:
Add rocprofv3_attach.py as an installed component

meserve-amd commented on projects/rocprofiler-sdk/source/bin/rocprofv3.py:
Promote -p from reserved to an advanced option that calls rocprofv3_attach.py
Also adds an option for duration `--attach-duration-msec`

meserve-amd commented on projects/rocprofiler-sdk/source/bin/rocprofv3_attach.py:
New file that loads and invokes the rocprofv3-attach library
Also contains the mechanics for detach (waits for enter, or uses duration)

meserve-amd commented on projects/rocprofiler-sdk/source/lib/CMakeLists.txt:
Add prestore subdirectory

meserve-amd commented on projects/rocprofiler-sdk/source/lib/output/generateCSV.cpp:
Prevents a crash when an agent_id handle was never recorded. This needs expansion of the prestore library to be fixed. I believe this change is good to keep regardless.

meserve-amd commented on projects/rocprofiler-sdk/source/lib/output/generateRocpd.cpp:
Prevents a crash when an agent_id handle was never recorded. This needs expansion of the prestore library to be fixed. I believe this change is good to keep regardless.

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/CMakeLists.txt:
Boilerplate cmake for the prestore library

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/code_object_registration.cpp:
This component stores all code_objects that have been frozen for later retrieval by rocprofiler-sdk
rocprofiler-register will indirectly call init, so we can modify the HSA table to capture these.
rocprofiler-sdk will call the `export_all` function to retrieve them
Only the handle needs to be saved, so the implementation here is fairly straightforward

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/code_object_registration.hpp:
"Public" header for the code object registration, used directly by rocprofiler-sdk

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/details/code_object_registration.hpp:
Private header for code object registration, contains class and type definitions

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/details/queue_registration.hpp:
Private header for queue registration, contains class and type definitions

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/prestore.cpp:
Prestore library main functions, called by rocprofiler-register at init

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/prestore.hpp:
Prestore public header

jrmadsen replied:
Public C headers should have .h extension

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/queue_registration.cpp:
This component stores all queues for later retrieval by rocprofiler-sdk
The implementation here is non-trivial due to two limitations:
- The queue must be initialized as an intercept queue
- The write interceptor function must be provided at creation time and cannot be changed

Thus, this code intercepts all create_queue calls to instead create intercept queues (same as rocprofiler-sdk). It records these as `queue_prestore_t` objects which contain the queue handle, agent, and user write interceptor function (initially nullptr).

It also implements a generic write interceptor function that by default is a passthrough. Later, rocprofiler-sdk can call the `set_write_interceptor` entry to update the queue_prestore entry for that queue, which will cause the generic write interceptor to call the provided write interceptor function instead.

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/queue_registration.hpp:
"Public" header for queue registration, used by rocprofiler-sdk

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/table.hpp:
Dispatch table format and versioning for prestore library, used by rocprofiler-sdk for casting
rocprofiler-register also contains this information, but not via this header

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp:
Attachment entry points used by rocprofiler-register to start the rocprofv3 tool when attaching

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/code_object/code_object.cpp:
Separates "executable_freeze" so that it can be called without doing a passthrough to HSA. This call is used when retrieving code objects from prestore. This process is facilitated by rocprofiler-register during the attach call after tool initialization.

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/code_object/code_object.hpp:
Add entrypoint for rocprofiler-register to trigger rocprofiler-sdk to load code objects from prestore

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/counters/metrics.cpp:
Resolves an issue when the invoked path is further inside our directory tree (e.g. not librocprofiler-sdk.so)

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/hip/stream.cpp:
HIP stream IDs are not saved in prestore currently, so this works around that

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/hsa/queue.cpp:
Adds a constructor for Queue which takes a queue which has already been created in HSA, such as those in queue_registration.
set_write_interceptor is a functor to provide the user's write interceptor function (usually created by QueueController)

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/hsa/queue.hpp:
Header changes for new Queue constructor

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/hsa/queue_controller.cpp:
Adds entrypoint for rocprofiler-register to call to retreive queues from prestore and provide the write interceptor functions needed.

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/hsa/queue_controller.hpp:
Add entrypoint for rocprofiler-register

meserve-amd commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/hsa/scratch_memory.cpp:
Downgrade this error message to keep everything running, agents need to be recorded in prestore and are not currently

meserve-amd commented on projects/rocprofiler-sdk/source/libexec/rocprofiler-sdk/CMakeLists.txt:
cmake to add attach component

meserve-amd commented on projects/rocprofiler-sdk/source/libexec/rocprofiler-sdk/rocprofiler-attach/CMakeLists.txt:
boilerplate cmake to add attach component

meserve-amd commented on projects/rocprofiler-sdk/source/libexec/rocprofiler-sdk/rocprofiler-attach/ptrace_session.cpp:
Large changes here:
- Ptrace support

ptrace is a system library used by gdb and others to allow control over other programs on the system.

We use a loop running in a separate thread to handle any signals. In ptrace, all signals are routed to the attacher first. We don't care about any of them, but we still need to pass them on.

We look up symbols in the target program by using our base address and symbol position to determine the symbol position from the base address in the target program. /proc contains information on the base address of libraries loaded in a process to make this possible.

We invoke functions to call mmap, munmap, and rocprofiler-register:attach
To invoke functions in the attachee, we do the following
- Stop execution
- Update the register file to invoke our function, including
    - Change PC to point to the program start address [1]
    - Change SP to not clobber redlined functions
    - Set parameters according to the x64 standard
- Write x64 assembly to the program start address to invoke the function and break
- Resume execution and wait for the break point
- Revert all changes and resume execution

[1] Program start address is used as it's the "least likely" location for a multi-threaded program to be executing. In a multi-threaded program, other threads will be allowed to continue executing, potentially executing our modified code and causing segfaults or worse.

meserve-amd commented on projects/rocprofiler-sdk/source/libexec/rocprofiler-sdk/rocprofiler-attach/ptrace_session.hpp:
Header for ptrace_session, used by attach main code

meserve-amd commented on projects/rocprofiler-sdk/source/libexec/rocprofiler-sdk/rocprofiler-attach/rocprofv3_attach.cpp:
Attachment main code. This component handles invoking ptrace and orchestrating the entire attach process. It also generates the environment buffer to be used by rocprofiler-register in the target process and ensures the buffer is available to the target process.

meserve-amd reviewed (commented):
File level comments to aid in review

jrmadsen commented on projects/rocprofiler-sdk/cmake/rocprofiler_config_packaging.cmake:
```diff
@@ -76,6 +76,7 @@ set(COMPONENT_NAME_docs "rocprofiler-sdk-docs")
 set(COMPONENT_NAME_tests "rocprofiler-sdk-tests")
 set(COMPONENT_NAME_roctx "rocprofiler-sdk-roctx")
 set(COMPONENT_NAME_rocpd "rocprofiler-sdk-rocpd")
+set(COMPONENT_NAME_prestore "rocprofiler-sdk-prestore")
```
Let’s name this component “attach” instead of “prestore”

jrmadsen replied:
This should be part of the normal rocprofiler-sdk package unless we create a `rocprofiler-sdk-tools` package

jrmadsen commented on projects/rocprofiler-sdk/source/bin/CMakeLists.txt:
```diff
@@ -26,6 +26,16 @@ install(
                 WORLD_EXECUTE
     COMPONENT tools)
 
+configure_file(rocprofv3_attach.py
```
```suggestion
configure_file(rocprofv3-attach.py
```
We use hyphens for executables, not underscores. Hyphens are far more common than underscores.

jrmadsen commented on projects/rocprofiler-register/source/lib/rocprofiler-register/rocprofiler_register.cpp:
```diff
@@ -601,6 +769,47 @@ rocprofiler_register_library_api_table(
         if(reginfo) (*reginfo)->propagated = true;
     }
+    else if(_import_match->library_idx == ROCP_REG_HSA &&
+            common::get_env("ROCPROFILER_REGISTER_ATTACHMENT_QUEUES_ENABLED", false))
```
So there needs to be a cmake flag that sets a ppdef that, when defined and > 0, sets this to true by default. This allows custom builds which always support attachment

jrmadsen commented on projects/rocprofiler-sdk/source/bin/rocprofv3_attach.py:
You need to support catching SIGINT (i.e. Ctrl+C) to stop data collection

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/prestore.cpp:
```diff
@@ -0,0 +1,76 @@
+void
+init_logging()
+{
+    rocprofiler::common::init_logging("ROCPROFILER");
```
```suggestion
    rocprofiler::common::init_logging("ROCPROFILER_ATTACH");
```

jrmadsen reviewed (changes requested):
Some first pass comments

meserve-amd committed:
attach: revert change to export level

ammarwa commented:
<details>
<summary><h2>Code Coverage Report</h2></summary>

<!-- Keep the empty line above to render markdown properly -->
# Code Coverage Report

## Tests Only

![code coverage tests.png](https://github.com/ROCm/rocm-systems/blob/bd9516e8743006f9ff2b2ec423f3499f7dc5f736/pr-316/tests.png?raw=True)

## Samples Only

![code coverage samples.png](https://github.com/ROCm/rocm-systems/blob/bd9516e8743006f9ff2b2ec423f3499f7dc5f736/pr-316/samples.png?raw=True)

## Tests + Samples

![code coverage all.png](https://github.com/ROCm/rocm-systems/blob/bd9516e8743006f9ff2b2ec423f3499f7dc5f736/pr-316/all.png?raw=True)

<!-- code-coverage-comment-identifier -->
</details>

tcgu-amd committed:
Make prestore API public

jrmadsen commented on projects/rocprofiler-sdk/source/bin/rocprofv3_attach.py:
Structure this to support use as an API, e.g. `def main(pid, attach_library=os.environ.get("ROCPROF_ATTACH_TOOL_LIBRARY", ROCPROF_ATTACH_TOOL_LIBRARY))` so that one, in theory, could import and call main from another python script

jrmadsen commented on projects/rocprofiler-sdk/source/bin/rocprofv3.py:
```diff
@@ -1420,6 +1429,16 @@ def log_config(_env):
+        update_env("ROCPROF_ATTACH_PID", args.pid)
+        if args.attach_duration_msec is not None:
+            update_env("ROCPROF_ATTACH_DURATION", f"{args.attach_duration_msec}")
+        path = os.path.join(f"{ROCM_DIR}", "bin/rocprofv3_attach")
```
```suggestion
        path = os.path.join(f"{ROCM_DIR}", "bin/rocprofv3-attach")
```

jrmadsen commented on projects/rocprofiler-sdk/source/lib/CMakeLists.txt:
```diff
@@ -8,6 +8,9 @@ add_subdirectory(common)
 add_subdirectory(rocprofiler-sdk)
 
+set(CMAKE_INSTALL_DEFAULT_COMPONENT_NAME "prestore")
+add_subdirectory(rocprofiler-sdk-prestore)
```
```suggestion
set(CMAKE_INSTALL_DEFAULT_COMPONENT_NAME "tools")
add_subdirectory(rocprofiler-sdk-attach)
```

jrmadsen commented on projects/rocprofiler-sdk/source/lib/output/generateCSV.cpp:
```diff
@@ -692,7 +692,10 @@ generate_csv(const output_config&
                 kind_name,
                 op_name,
-                tool_metadata.get_agent_index(record.agent_id, cfg.agent_index_value).as_string(),
+                record.agent_id.handle != 0
+                    ? tool_metadata.get_agent_index(record.agent_id, cfg.agent_index_value)
+                          .as_string()
+                    : "UNK",  // Some agents may be unkown due to rocprofiler attachment feature
```
We should not have unknown agents because of attachment

jrmadsen replied:
@meserve-amd please see above comment

jrmadsen commented on projects/rocprofiler-sdk/source/lib/output/generateRocpd.cpp:
```diff
@@ -1226,7 +1226,10 @@ write_rocpd(
                     if(_type == "ALLOC")
                     {
-                        _node_id = tool_metadata.get_agent(itr.agent_id)->node_id;
+                        // Some agents for scratch memory may be unkown due to rocprofiler
+                        // attachment feature
+                        auto* agent = tool_metadata.get_agent(itr.agent_id);
+                        _node_id    = agent != nullptr ? agent->node_id : 0;
```
This change shouldn't be necessary

meserve-amd replied:
You're right, this is a vestige of an earlier change that did not handle the queues' agents and this _should_ work now. I will try without and revert it.

jrmadsen replied:
@meserve-amd please see original comment

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/CMakeLists.txt:
`prestore` -> `attach`

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/code_object_registration.cpp:
```diff
@@ -0,0 +1,135 @@
+ROCPROFILER_EXTERN_C_INIT
+
+int
+rocprofiler_prestore_export_all_code_objects(hsa_executable_t* executables,
```
make this an "iterate" function, e.g. rocprofiler-sdk invokes it with a callback and the callback gets invoked for each executable.

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/code_object_registration.hpp:
```diff
@@ -0,0 +1,48 @@
+#include <cstdint>
+
+ROCPROFILER_EXTERN_C_INIT
+
```
Extern C functions at the end of the file.

meserve-amd replied:
Will update this across all files

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/details/code_object_registration.hpp:
```diff
@@ -0,0 +1,75 @@
+    code_object_collection_t m_code_objects;
+
+    hsa_executable_freeze_t  m_hsa_executable_freeze_fn;
+    hsa_executable_destroy_t m_hsa_executable_destroy_fn;
```
Need to be initialized

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/details/code_object_registration.hpp:
```diff
@@ -0,0 +1,75 @@
+
+    hsa_executable_freeze_t  m_hsa_executable_freeze_fn;
+    hsa_executable_destroy_t m_hsa_executable_destroy_fn;
+};
```
Prefer a functional-style with the class being part of the implementation details.

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/details/queue_registration.hpp:
```diff
@@ -0,0 +1,92 @@
+    hsa_queue_t*        queue;
+    write_interceptor_t user_write_interceptor_func;
+    void*               user_write_interceptor_data;
+};
```
Add default initializers

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/details/queue_registration.hpp:
```diff
@@ -0,0 +1,92 @@
+             uint32_t           private_segment_size,
+             uint32_t           group_segment_size,
+             CoreApiTable       core_api,
+             AmdExtTable        ext_api);
```
Do not forward declare functions in a header that are not called outside of the implementation

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/details/queue_registration.hpp:
```diff
@@ -0,0 +1,92 @@
+    AmdExtTable  m_ext_table  = {};
+
+    queue_collection_t m_queues;
+};
```
Prefer a functional style over OOP

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/details/queue_registration.hpp:
```diff
@@ -0,0 +1,92 @@
+
+    queue_collection_t& get_all_queues() { return m_queues; };
+    const CoreApiTable& get_core_table() const { return m_core_table; }
+    const AmdExtTable&  get_ext_table() const { return m_ext_table; }
```
Who needs this?

bwelton replied:
I second this, there should only be one location for these tables really.

meserve-amd replied:
This mirrors how queue_controller handles these tables now. At a minimum, we need to at least keep a reference so that our create_queue wrapper can make an intercept queue in HSA during runtime. It's public so that the wrapper can call into the table directly, but I can hide it in the class instead.

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/queue_registration.cpp:
```diff
@@ -0,0 +1,225 @@
+                  hsa_amd_queue_intercept_packet_writer_t writer)
+{
+    ROCP_FATAL_IF(data == nullptr) << "WriteInterceptor was not passed a pointer to the queue";
+    auto queue = static_cast<hsa_queue_t*>(data);
```
Use a struct with `hsa_queue_t` and an atomic `user_write_interceptor_fn` and whatever else you need here. You shouldn't need to use any global "queue registration" class here

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/queue_registration.cpp:
```diff
@@ -0,0 +1,225 @@
+        FATAL, ext_api.hsa_amd_queue_intercept_register_fn(queue, write_interceptor, queue))
+        << "Could not register interceptor";
+
+    queue_prestore_t queue_prestore{};
+    queue_prestore.agent                       = agent;
+    queue_prestore.queue                       = queue;
```
This should be storing all the arguments (i.e. `size`, `type`, etc.)

meserve-amd replied:
My thinking here was we don't use these parameters again once the queue is created in HSA. We can save them, but they aren't used anywhere else. This is also true in the normal codeflow of rocprofiler-sdk.

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/queue_registration.cpp:
```diff
@@ -0,0 +1,225 @@
+}
+
+queue_prestore_t
+create_queue_prestore(hsa_agent_t        agent,
```
This looks like a function that should be in an anonymous namespace

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/queue_registration.cpp:
```diff
@@ -0,0 +1,225 @@
+// This shim will look up the associated queue and execute the user's write intercept function
+// if it has been previously provided. Otherwise, it will execute the packet write normally.
+void
+write_interceptor(const void*                             packets,
```
anonymous namespace

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/queue_registration.cpp:
```diff
@@ -0,0 +1,225 @@
+void
+QueueRegistration::add_queue(queue_prestore_t queue_prestore)
+{
+    m_queues[queue_prestore.queue] = queue_prestore;
```
this overwrites. are you sure that you want unconditional overwriting?

meserve-amd replied:
The handles we get _should_ be unique, so I'll log an error here at least

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/queue_registration.cpp:
```diff
@@ -0,0 +1,225 @@
+ROCPROFILER_EXTERN_C_INIT
+
+int
+rocprofiler_prestore_export_all_queues(rocprofiler::prestore::queue_prestore_export_t* queues,
```
you should not have an "extern C" function with C++ parameters

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/queue_registration.hpp:
```diff
@@ -0,0 +1,65 @@
+
+int
+rocprofiler_prestore_export_all_queues(rocprofiler::prestore::queue_prestore_export_t* queues,
+                                       uint64_t* num_queues) ROCPROFILER_PUBLIC_API;
```
Why are these public API functions when everything is supposed to go through a dispatch table?

meserve-amd replied:
This is a byproduct of register.so creating the dispatch table for this library, which will go away when that is fixed

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-prestore/table.hpp:
Make this a C compatible header

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/code_object/code_object.cpp:
```diff
@@ -1219,3 +1229,43 @@ iterate_loaded_code_objects(code_object_iterator_t&& func)
+ROCPROFILER_EXTERN_C_INIT
+
+int
+rocprofiler_load_prestore_code_objects(void* incoming_table)
```
Who is calling this externally?

meserve-amd replied:
This is called by register post-init to signal rocprofiler-sdk that it should go and get the saved code objects. This will go away and be called when the attach.so table is registered with rocprofiler-sdk

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/code_object/code_object.hpp:
```diff
@@ -31,6 +31,13 @@
+// Hidden function used to load all previously captured code objects after an attachment.
+// Takes a dispatch table of prestore functions usually provided by rocprofiler_register.
+int
+rocprofiler_load_prestore_code_objects(void* incoming_table) ROCPROFILER_PUBLIC_API;
```
Why is this public API?

meserve-amd replied:
See [here](https://github.com/ROCm/rocm-systems/pull/316#discussion_r2285901802), this will go away

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/counters/metrics.cpp:
```diff
@@ -236,7 +236,18 @@ findViaInstallPath(const std::string& filename)
+                   .parent_path()
+                   .parent_path()
+                   .parent_path()
+                   .parent_path() /
```
What is going on here?

meserve-amd replied:
This was a quick fix for a case where the install path was a tools library, but I don't think that's needed anymore. I will investigate and clean this up

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/hip/stream.cpp:
```diff
@@ -108,11 +108,14 @@ add_stream(hipStream_t stream)
 get_stream_id(hipStream_t stream)
 {
+    // The check for stream is a hack to make this work. You should not use rocprofv3 attach without
+    // --group-by-queue
```
This seems like a restriction we need to fix.

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/hsa/queue_controller.cpp:
```diff
@@ -482,3 +484,77 @@ queue_controller_fini()
+ROCPROFILER_EXTERN_C_INIT
+
+int
+rocprofiler_load_prestore_queues(void* incoming_table)
```
This should not be called from rocprofiler-register

jrmadsen reviewed (changes requested):
The design is not quite what I recall us discussing. rocprofiler-register should dlopen `rocprofiler-sdk-attach` and `rocprofiler-sdk-attach` should register a dispatch table with rocprofiler-register. When rocprofv3 attaches, it should trigger rocprofiler-register passing the `rocprofiler-sdk-attach` dispatch table to it.

bwelton commented on projects/rocprofiler-sdk/source/bin/rocprofv3_attach.py:
```diff
@@ -0,0 +1,64 @@
+libname = os.environ.get("ROCPROF_ATTACH_TOOL_LIBRARY", ROCPROF_ATTACH_TOOL_LIBRARY)
+c_lib = ctypes.CDLL(libname)
+
+if c_lib is None:
```
ctypes raises an error on failure, doesn't return none.

bwelton commented on projects/rocprofiler-sdk/source/libexec/rocprofiler-sdk/rocprofiler-attach/ptrace_session.cpp:
```diff
@@ -0,0 +1,701 @@
+    // Load the library in our process to determine the offset of the requested symbol from the
+    // start address of the library
+    addr        = nullptr;
+    libraryaddr = dlopen(library.c_str(), RTLD_LAZY);
```
rocp_load_lib?

meserve-amd replied:
You're right, but that's in register, so I don't have access to that function here (this is on attacher side too). I will copy it over and we can move it to common later.

bwelton commented on projects/rocprofiler-register/source/lib/rocprofiler-register/rocprofiler_register.cpp:
```diff
@@ -639,25 +848,161 @@ rocprofiler_register_iterate_registration_info(
+rocprofiler_register_detach() ROCPROFILER_REGISTER_PUBLIC_API;
+
+//
+//  This function can be invoked by ptrace
```
What does this mean? That at some arbitrary point in program execution this can be called? 

The primary concern with this PR that i have is that its not clear to me how to check what the state of attachment we are in. This is really a state machine with ~5 positions (detached, attached, attaching, detaching, failed). It seems pretty difficult to determine what state we are in from any particular call. This makes error handling, testing, and understanding of this feature difficult. 

What I propose is to have a manger struct that tracks the overall state + contains the data structures being operated on.  Something like this:

struct AttachmentState {
      AttachmentState state_;
      pid_t target_pid_;
      std::unique_ptr<EnvironmentManager> env_mgr_;
      std::unique_ptr<LibraryManager> lib_mgr_;
      std::unique_ptr<PtraceInterface> ptrace_;
}; 

Then have functions that operate on this struct to centralize the data being used for attachment. This would also help out massively on the other concern which is a lack of testing (since you can now "mock" the state of the connection/environment/etc).

meserve-amd replied:
Yes, this code is on the "attachee" side, so we expect this function be called at some arbitrary point by ptrace. This is carried out by a thread hijacked by ptrace. I wrote [a file level comment in ptrace_session.cpp](https://github.com/ROCm/rocm-systems/pull/316#discussion_r2270800207) to explain this process, but GitHub helpfully marked that comment as outdated despite that file not changing. I will elaborate a little further about the expected code flow in the comment here. It needs to be public to ensure the symbol can be found and executed.

I am assuming your comment about state tracking is more about the libexec/rocprofiler-sdk/rocprofiler-attach portion of the code. The PTraceSession class does some of this tracking across some of its booleans, but I think you're right that it needs to be more centralized and some of the disparate parts can be separated out to make testing easier and then used together e.g. PTraceSession doesn't need to do the library load and symbol calculation, a separate module can do that and give the target address.

bwelton reviewed (changes requested):
Most of my comments are conceptual in nature in that i believe that we need to have a more clear design here.

meserve-amd committed:
attach: update sdk attachment library WIP

- This commit is NONFUNCTIONAL

- Changes around structure to remove classes
- Seperate C linkage where needed
- Still needs updates to register for correct usage

meserve-amd committed:
attach: update register with dispatch table WIP
- This commit is NONFUNCTIONAL

- Changes rocprofiler_register to handle dispatch table from attach
  library.
- Still needs changes in SDK with dispatch table usage

meserve-amd committed:
attach: dispatch table wip
- This commit is NONFUNCTIONAL

meserve-amd committed:
attach: move attach component into core

meserve-amd committed:
attach: rename to rocprofv3-attach

meserve-amd committed:
attach: add callbacks for new queues and code objects

meserve-amd committed:
attach: finish dispatch table implementation

- Fixes kernel tracing

claude committed:
feat: Add --attach alias for rocprofv3 with comprehensive attachment tests

- Add `--attach` as an alias to existing `-p/--pid` functionality in rocprofv3.py
- Create comprehensive attachment test suite with CSV and JSON output validation:
- New attachment-test application for testing dynamic profiling scenarios
- Unified test script supporting both CSV and JSON output formats
- Pytest-based validation for kernel traces, memory copies, HSA API calls, and agent info
- Add CMake integration for automated attachment testing
- Support parameterized output directory and filename specification
- Implement proper environment setup for attachment queue registration

Tests verify successful attachment to running processes and capture of:
- Kernel dispatch traces with workgroup/grid dimensions
- Memory copy operations (H2D/D2H) with size validation
- HSA API call traces across multiple domains
- GPU/CPU agent information and capabilities

meserve-amd committed:
attach: add cmake variable for attachment support

Commit:
Documentation Update

meserve-amd committed:
attach: make attach script callable

Commit:
Added ROCPROFILER_REGISTER_ATTACHMENT_TOOL_LIB to remove hardcoded name

meserve-amd committed:
attach: revert metrics library path changes

itrowbri committed:
Generic Attachment in Register (#942)

Remove tool references in register

Commit:
Add second param to attach call in rocprof register

Commit:
Add experimental reattachment support for ROCprofiler-SDK

This commit introduces experimental reattachment functionality allowing tools
to dynamically reattach to running processes with comprehensive design changes
to support multiple attach/detach cycles:

**Core Reattachment API:**
- Add rocprofiler_tool_configure_result_experimental_t with tool_reattach/tool_detach callbacks
- Add rocprofiler_call_client_reattach and rocprofiler_call_client_detach C exports
- Implement reattachment tracking in rocprofiler_register_attach to differentiate
initial attachment from reattachment cycles
- Add rocprofiler_register_invoke_reattach for handling reattachment requests

**Design Changes - Registration System Flow:**
The registration system now supports a dual-path initialization:

1. Initial Attachment Flow:
    - rocprofiler_register_attach() -> rocprofiler_register_invoke_all_registrations()
    - Full tool initialization with complete context setup
    - Sets prev_attached atomic flag to track state

2. Reattachment Flow:
    - rocprofiler_register_attach() detects prev_attached=true -> rocprofiler_register_invoke_reattach()
    - Bypasses full re-initialization, calls client reattach callbacks instead
    - Preserves existing contexts and buffers, only reactivates profiling services

**Design Changes - Tool Library Loading:**
Enhanced rocprofiler-register library loading with function pointer resolution:
- Extended rocp_set_api_table_data_t tuple to include reattach/detach function pointers
- Automatic symbol resolution for rocprofiler_call_client_reattach/detach functions
- Support for both LD_PRELOAD and dlopen scenarios with consistent callback availability

**Design Changes - Context Management:**
Introduced dual context systems for attachment scenarios:
- get_contexts() - Original contexts for standard tool initialization
- get_attach_contexts() - Separate context map for attachment-specific lifecycle
- attach_init() - Creates contexts for ALL buffer tracing services using existing buffers
- attach_start() - Selectively starts contexts based on configuration options
- attach_detach() - Cleanly stops and destroys attachment contexts

**Design Changes - Buffer Management:**
Added reset_tmp_file_buffer() template for clean reattachment state:
- Properly closes and removes old temporary files
- Deletes existing file_buffer instances to prevent stale file position tracking
- Creates fresh file_buffer instances for clean reattachment cycles
- Addresses core issue where file position metadata becomes stale between cycles

**Design Changes - Environment Variable Injection:**
Added ROCP_REGISTERED_TOOL_ATTACH environment variable:
- Distinguishes attachment-loaded tools from LD_PRELOAD scenarios
- Enables registration system to apply attachment-specific logic
- Helps tools adapt behavior for attachment vs standard initialization

**Attachment Context Management:**
- Add attach_init/attach_start/attach_detach functions for dynamic context lifecycle
- Add reset_tmp_file_buffer template for clean reattachment state management
- Implement get_attach_contexts() for tracking active attachment contexts

**Test Infrastructure:**
- Add projects/rocprofiler-sdk/tests/rocprofv3/reattach/ comprehensive test suite
- Include reattachment test scripts with unified attachment/detachment cycles
- Add validate.py with trace data validation for kernel, memory copy, HSA API, and agent info
- Add conftest.py for JSON and CSV data loading utilities

**Configuration Updates:**
- Update CMakeLists.txt to include reattachment tests in build system
- Add environment variable ROCP_REGISTERED_TOOL_ATTACH for attachment state tracking
- Enhance rocprofiler-register library loading with reattach/detach function resolution

**Flow Impact Analysis:**
This design enables robust multi-cycle attachment by:
1. Preventing duplicate initialization on reattachment
2. Maintaining separate context lifecycles for attachment vs standard operation
3. Ensuring clean temporary file state between attachment cycles
4. Providing tools with explicit reattach/detach callback hooks
5. Supporting both programmatic and environment-based tool configuration

The experimental nature allows for iteration on the API while establishing
the foundation for production-ready dynamic profiling capabilities.

jrmadsen committed:
Fix misc clang-tidy warnings/errors

jrmadsen committed:
CMake Option and Environment Variable Updates

- CMake: ROCPROFILER_REGISTER_ALWAYS_SUPPORT_ATTACH -> ROCPROFILER_REGISTER_BUILD_DEFAULT_ATTACHMENT
- Env: ROCPROFILER_REGISTER_ATTACHMENT_ENABLED ->

jrmadsen committed:
Source reorganization

jrmadsen committed:
Formatting + new lines at EOF

jrmadsen committed:
Fix flake8 F841: local variable is assigned to but never used

jrmadsen committed:
Update attachment test

- get rid of 5 second start delay
- add roctx

jrmadsen committed:
Rework implementation

- Remove rocprofiler_tool_configure_result_experimental_t in lieu of rocprofiler_configure_attach
- Add <rocprofiler-sdk/experimental/registration.h>
- TODO: Update process_attachment.rst

jrmadsen commented on projects/rocprofiler-sdk/source/bin/rocprofv3-attach.py:
```diff
@@ -0,0 +1,65 @@
+    c_lib = ctypes.CDLL(attach_library)
+
+    c_lib.attach.argtypes = [ctypes.c_uint]
+    c_lib.attach(int(pid))
```
We need error handling here

jrmadsen commented on projects/rocprofiler-sdk/source/bin/rocprofv3.py:
```diff
@@ -1298,6 +1305,13 @@ def _write_env_value():
             overwrite_if_true=True,
         )
 
+    if args.pid:
```
This was necessary to prevent LD_PRELOAD on rocprofv3-attach

jrmadsen commented on projects/rocprofiler-sdk/source/docs/api-reference/process_attachment.rst:
This file needs to be updated

jrmadsen commented on projects/rocprofiler-register/source/include/rocprofiler-register/rocprofiler-register.h:
```diff
@@ -152,6 +156,17 @@ rocprofiler_register_iterate_registration_info(
     void*                                       data)
     ROCPROFILER_REGISTER_ATTRIBUTE(nonnull(1)) ROCPROFILER_REGISTER_PUBLIC_API;
 
+/// @brief Function for attaching to a process and loading tool library
+/// @param[in] environment_buffer Environment variables to set
+/// @param[in] tool_lib_path Path to the tool library to load (can be NULL)
+/// @return ::rocprofiler_register_error_code_t
+rocprofiler_register_error_code_t
+rocprofiler_register_attach(const char* environment_buffer, const char* tool_lib_path)
+    ROCPROFILER_REGISTER_ATTRIBUTE(nonnull(2)) ROCPROFILER_REGISTER_PUBLIC_API;
+
+rocprofiler_register_error_code_t
+rocprofiler_register_detach() ROCPROFILER_REGISTER_PUBLIC_API;
```
This probably should not be declared in the header while we are experimental

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-attach/attach.cpp:
```diff
@@ -0,0 +1,90 @@
+                                 uint64_t    lib_instance,
+                                 void**      tables,
+                                 uint64_t    num_tables,
+                                 void*       register_functor)
```
I would prefer the `register_functor` to be `decltype(::rocprofiler_set_api_table)*` -- i.e. function pointer to function with signature identical to `rocprofiler_set_api_table`

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-attach/attach.cpp:
```diff
@@ -0,0 +1,90 @@
+    return ROCPROFILER_STATUS_SUCCESS;
+}
+
+int
+rocprofiler_attach_get_version()
+{
+    constexpr int ROCPROFILER_ATTACH_VERSION = 1;
+    return ROCPROFILER_ATTACH_VERSION;
+}
```
Why only a single version number? This needs to be fixed to be the same form as `ROCPROFILER_VERSION` which encodes the major, minor, and patch

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/README.md:
```diff
@@ -10,7 +10,9 @@ with the rocprofiler.
-services from the rocprofiler.
+services from the rocprofiler. rocprofv3 supports both
+launching new applications and attaching to existing
+processes using the --attach/--pid/-p options.
```
```suggestion
processes using the `--attach`/`--pid`/`-p` options.
```

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp:
```diff
@@ -1767,6 +1769,36 @@ get_tracing_callbacks()
+    // reset config for attach (i.e. re-parse environment variables)
+    rocprofiler::tool::get_config() = rocprofiler::tool::config{};
+
+    pid_t target_pid = getppid();  // The target process we're attaching to
+    pid_t tool_pid   = getpid();   // The rocprofv3 tool process
+    ROCP_INFO << "Attach mode: Setting process_id to target PID " << target_pid
+              << " (tool PID: " << tool_pid << ")";
+    tool_metadata->set_process_id(target_pid, 0);  // Set target as main process
```
@bwelton is this necessary?

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp:
```diff
@@ -3072,13 +3271,54 @@ rocprofiler_configure(uint32_t                 version,
               << " (" << runtime_version << ")";
 
-    // create configure data
+    // create configure data using experimental struct with attach/detach support
```
```suggestion
    // create configure data
```

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp:
```diff
@@ -3072,13 +3271,54 @@ rocprofiler_configure(uint32_t                 version,
+                             uint32_t                 priority,
+                             rocprofiler_client_id_t* id)
+{
+    initialize_logging();
+
+    // set the client name
+    id->name = "rocprofv3";
+
+    // store client info
+    client_identifier = id;
+
+    // note that rocprofv3 is not the primary tool
+    ROCP_WARNING_IF(priority > 0) << id->name << " has a priority of " << priority
+                                  << " (not primary tool)";
+
+    // compute major/minor/patch version info
+    uint32_t major = version / 10000;
+    uint32_t minor = (version % 10000) / 100;
+    uint32_t patch = version % 100;
+
+    // ensure these pointers are not leaked
+    add_destructor(tool_metadata);
+    add_destructor(execution_profile);
+
+    tool::get_tmp_file_name_callback() = [](domain_type type) -> std::string {
+        return compose_tmp_file_name(tool::get_config(), type);
+    };
+
+    ROCP_INFO << id->name << " is using rocprofiler-sdk v" << major << "." << minor << "." << patch
+              << " (" << runtime_version << ")";
+
```
This can probably be removed.

jrmadsen reviewed (changes requested):
Resetting the config in the `tool_attach` allows some env variables to be changed on reattachment (i.e. output paths) but others are ignored (i.e. services) -- we need error handling on service changes that we do not currently support.

jrmadsen committed:
Handle re-attachment options

- inherit options from previous attachment
- check previous options do not modify data collection services

jrmadsen committed:
Fix support for tools w/o rocprofiler_configure_attach

- fix segfault when rocprofiler_configure_attach does not exist
- fix naming convention for functions accepting attach dispatch table
- cleanup rocprofiler_configure_attach implementation in rocprofv3 tool

meserve-amd committed:
attach: remove unknown agent handling

- Change was from earlier commit, no longer needed

meserve-amd committed:
attach: add error for attaching without library loaded

meserve-amd committed:
attach: revise version numbering

meserve-amd committed:
attach: register header revisions

meserve-amd committed:
attach: clang format register

jrmadsen commented on projects/rocprofiler-register/source/lib/rocprofiler-register/rocprofiler_register.cpp:
```diff
@@ -634,6 +637,22 @@ load_environment_buffer(const char* environment_buffer)
+            break;
+        }
+    }
+    return _attachment_library_registered;
```
```suggestion
    for (const auto& itr : registered)
    {
        if (std::string_view{itr->common_name} == 
            supported_library_trait<ROCP_REG_ROCATTACH>::common_name)
            return true;
    }
    return false;
```

meserve-amd committed:
attach: formatting

meserve-amd committed:
attach: fix build failure

- Remove cross dependency into rocprofiler-sdk, fixes build on some systems

meserve-amd committed:
attach: revise register library detection

jrmadsen committed:
Update rocprofiler-register and attach library

- formatting
- proper signature of register_functor for rocprofiler-sdk-attach library callback
- remove get_dispatch_registration_table()

jrmadsen committed:
Bump rocprofiler-register version to 0.6.0 + AnyNewerVersion

jrmadsen committed:
Fix output support for rocprofiler-sdk-tool

jrmadsen committed:
Fix formatting

jrmadsen committed:
Fix clang tidy errors

jrmadsen committed:
Misc rocprofiler-sdk-attach fixes

meserve-amd committed:
attach: add sigint handling to attach python

meserve-amd committed:
tool README.md formatting

Co-authored-by: Jonathan R. Madsen <jrmadsen@users.noreply.github.com>

jrmadsen committed:
Fix buffered output issue

meserve-amd committed:
attach: add errors for tool attach

bwelton committed:
CI Fixes

bwelton commented on projects/rocprofiler-sdk/tests/rocprofv3/attachment/validate.py:
```diff
@@ -111,10 +111,10 @@ def test_attachment_hsa_api_trace(hsa_input_data):
-    assert any(
-        "memory" in func.lower() for func in functions
-    ), "No memory-related HSA functions captured"
+    # Note: Memory-related HSA function checking disabled for RHEL/SLES compatibility
```
This is primarily to do with the attachment timing.

jrmadsen committed:
Rework tests

meserve-amd committed:
attach: improve library loading in rocprofv3 attach

jrmadsen committed:
formatting

jrmadsen committed:
Update tests to use pytest framework

jrmadsen committed:
Fix test_attachment_hsa_api_trace

meserve-amd committed:
attach: catch ctypes exceptions

meserve-amd committed:
attach: fix leak in registration

meserve-amd committed:
attach: fix sanitizer tests

meserve-amd committed:
attach: fix sanitizer tests further

meserve-amd committed:
attach: disable attach asan tests

meserve-amd committed:
attach: disable ubsan test

meserve-amd committed:
attach: fix permissions in installed test package

meserve-amd committed:
attach: formatting
